File: java/n_queens_ii.java
public class Solution {
    int sum;
    
    boolean isValid(int[] pos, int n, int start, int col) {
        for (int j=0; j<start; j++) {
            if (pos[j]==col || Math.abs(start-j)==Math.abs(col-pos[j])) {
                return false;
            }
        }
        return true;
    }
    
    void getSum(int[] pos, int n, int start) {
        if (start==n) {
            sum++;
            return;
        }
        
        for (int col=0; col<n; col++) {
            if (isValid(pos, n, start, col)){
                pos[start]=col;
                getSum(pos, n, start+1);
            }
        }
    }
    
    public int totalNQueens(int n) {
        sum=0;
        if (n<=0) return 0;
        int[] pos=new int[n];
        getSum(pos, n, 0);
        return sum;
    }
}

File: cpp/n-queens-ii.cpp
52. N-Queens II
Follow up for N-Queens problem.

Now, instead outputting board configurations, return the total number of distinct solutions.

Hide Company Tags Zenefits
Hide Tags Backtracking
Hide Similar Problems (H) N-Queens

class Solution {
public:
    int totalNQueens(int n) {
        int res=0;
        vector<int> pos;
        dfs(res, pos, n, 0);

        return res;
    }
private:
    void dfs(int& res, vector<int>& pos, int n, int start) {
        if (start==n) {
            res++;
            return;
        }

        for (int col=0; col<n; ++col) {
            if (isValid(pos, col)) {
                pos.push_back(col);
                dfs(res, pos, n, start+1);
                pos.pop_back();
            }
        }
    }

    bool isValid(vector<int>& pos, int col) {
        int row=pos.size();
        for (int i=0; i<pos.size(); ++i) {
            if (pos[i]==col || abs(pos[i]-col)==row-i) return false;
        }
        return true;
    }
};
